### DISCUSSION TITLE
Best practice for implementing list/detail views for complex structures

### DISCUSSION THREAD

**Discussion by nareshbhatia** [graphql-rules/graphql-rules#120] [Q&A] [answered]
I am building a list/detail view for a complex alerting system. Items in the `AlertList` must show high level summaries of alerts. When an item in the list is clicked, the detail view must show all the details of the alert.

Since the Alert object is very complex, the query to fetch the `AlertList` is fairly deep. It must reach out to different attributes of different alert types to create a summary. In such cases, is it a good practice to create a new GraphQL type called `AlertSummary` which is computed by the server just to make displaying the `AlertList` easier. `AlertSummary` will essentially be a subset of the the full `Alert` entity. Is this a good practice or an anti-pattern? Is it covered by any rule in this project?

**Comment by RIP21** [comment, marked as answer]
You can make it so it's a field of the `Alert` GraphQL type.
Like `summary` or so. 
So regular `Alert` root fields will have only raw fields (filtered from those that UI don't need, exposing DB types 1:1 to GraphQL schema is a bad practice, or even dangerous security-wise, plus every exposed field is hard to take away without breaking changes, so each exposed field should be considered)

Smth like
```graphql
type Alert implements Node {
  id: ID!
  # ... other fields
  summary: AlertSummary!
}

type AlertSummary {
  # ... summary fields
}
```

This way you'll be able to implement a `summary` field resolver that will only be called in case this summary is requested.
If it's not needed, then this, potentially heavy operation, will not be used, hence server load will be lower and response times faster.

Just make sure the web or any other client uses fragment colocation and other techniques to ensure no over-fetching (these are the default for Relay)

This can be achieved in other clients (if we talking TS) with this approach:
https://the-guild.dev/blog/unleash-the-power-of-fragments-with-graphql-codegen

**Reply by nareshbhatia** [reply]
Thank you @RIP21 – good guidance.

Just one question, what is `Node` in your suggested schema? Is this a suggested pattern/best practice somewhere?

**Reply by RIP21** [reply]
@nareshbhatia you can read about Relay Spec and what GraphQL API considered Relay compatible here.
https://relay.dev/docs/guides/graphql-server-specification/

You can also read why Relay is great here:
https://alan.norbauer.com/articles/relay-style-graphql

**Reply by nareshbhatia** [reply]
Thank you, @RIP21. Will be reading both.
